kjdoc/tour.md

```diff
@@ -538,13 +538,441 @@ This section describes KJ APIs that control process execution and low-level inte
+```
+
+### Threads
+
+The KJ async framework is designed around single-threaded event loops. However, you can multiple threads, with each running its own loop.
+
+All KJ async objects, unless specifically documented otherwise, are intrinsically tied to the thread and event loop on which they were created. These objects must not be accessed from any other thread.
+
+To communicate between threads, you may use `kj::Executor`. Each thread (that has an event loop) may call `kj::getCurrentThreadExecutor()` to get a reference to its own `Executor`. That reference may then be shared with other threads. The other threads can use the methods of `Executor` to queue functions to execute on the owning thread's event loop.
+
+The threads which call an `Executor` do not have to have KJ event loops themselves. Thus, you can use an `Executor` to signal a KJ event loop thread from a non-KJ thread.
+
+### Fibers
+
+Fibers allow code to be written in a synchronous / blocking style while running inside the KJ event loop, by executing the code on an alternate call stack. The code running on this alternate stack is given a special `kj::WaitScope&`, which it can pass to `promise.wait()` to perform synchronous waits. When such a `.wait()` is invoked, the thread switches back to the main call stack and continues running the event loop there. When the waited promise resolves, execution switches back to the alternate call stack and `.wait()` returns (or throws).
+
+```c++
+constexpr size_t STACK_SIZE = 65536;
+kj::Promise<int> promise =
+    kj::startFiber(STACK_SIZE, [](kj::WaitScope& waitScope) {
+  int i = someAsyncFunc().wait(waitScope);
+  i += anotherAsyncFunc().wait(waitScope);
+  return i;
+});
+```
+
+**CAUTION:** Fibers produce attractive-looking code, but have serious drawbacks. Every fiber must allocate a new call stack, which is typically rather large. The above example allocates a 64kb stack, which is the _minimum_ supported size. Some programs and libraries expect to be able to allocate megabytes of data on the stack. On modern Linux systems, a default stack size of 8MB is typical. Stacks space is allocated lazily on page faults, but just setting up the memory mapping is much more expensive than a typical `malloc()`. If you create lots of fibers, you should use `kj::FiberPool` to reduce allocation costs -- but while this reduces allocation overhead, it will increase memory usage.
+
+Because of this, fibers should not be used just to make code look nice (C++20's `co_await`, which KJ will soon support, is a better way to do that). Instead, the main use case for fibers is to be able to call into existing libraries that are not designed to operate in an asynchronous way. For example, say you find a library that performs stream I/O, and lets you provide your own `read()`/`write()` implementations, but expects those implementations to operate in a blocking fashion. With fibers, you can use such a library within the asynchronous KJ event loop.
```

Comment:
```suggestion
Because of this, fibers should not be used just to make code look nice (C++20's `co_await`, which KJ will soon support, is a better way to do that.) Instead, the main use case for fibers is to be able to call into existing libraries that are not designed to operate in an asynchronous way. For example, say you find a library that performs stream I/O, and lets you provide your own `read()`/`write()` implementations, but expects those implementations to operate in a blocking fashion. With fibers, you can use such a library within the asynchronous KJ event loop.
```

Reply:
Hmm, really? I know that it's said that when a sentence ends with a quoted phrase, the period is supposed to be inside the quote. As a programmer, I can't deal with that rule and usually ignore it. But I thought that when a sentence ends with a parenthetical, it's normal for the period to be outside the parenthetical?

Or are you suggesting that the parenthetical should be separated from the preceding sentence entirely? In that case I suppose we need to add a period before the parenthetical?

Reply:
The answers to all four of your questions is: I made a mistake. :(

Reply:
Haha well you're the linguist so I trusted you!